### PR TITLE
Bluetooth: GATT: Add support to setting permission on CCCD

### LINKED ISSUE #17983: Bluetooth: Re-establish security before notifications/indications can be sent
**EDIT**: This issue has been hijacked. The original issue was actually a problem on the applicaiton side, but we reuse this one to cover what is described in [this comment](https://github.com/zephyrproject-rtos/zephyr/issues/17983#issuecomment-526572654)

-------
Old description

**Describe the bug**
See `notify_connected` in `subsys/bluetooth/host/conn.c`.
When all callbacks are executed gatt is notified. It seems OK except that it leads to data exchange before security is established (we get information about service's notification).

I seen it while trying to move the code where we sent data for the first time in the app. Can be worked around but maybe should be fixed some other way.

**To Reproduce**
N/A

**Expected behavior**
I expected no exchange before security is set.

**Impact**
?

**Screenshots or console output**
N/A

**Environment (please complete the following information):**
master

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by Vudentz**
This adds support to set different permissions to CCCD so security can
be checked when enabling notification which conforms to:

BLUETOOTH CORE SPECIFICATION Version 5.1 | Vol 3, Part G page 2360:

  '3.3.3.3 Client Characteristic Configuration

   Authentication and authorization may be required by the server to
   write the configuration descriptor.'

In addition to that also ensure that notification are not re-enabled
until the proper security level is reached to conform to the following
statement:

  '10.3.1.1 Handling of GATT indications and notifications

   A client “requests” a server to send indications and notifications
   by appropriately configuring the server via a Client Characteristic
   Configuration Descriptor. Since the configuration is persistent
   across a disconnection and reconnection, security requirements must
   be checked against the configuration upon a reconnection before
   sending indications or notifications. When a server reconnects to a
   client to send an indication or notification for which security is
   required, the server shall initiate or request encryption with the
   client prior to sending an indication or notification. If the client
   does not have an LTK indicating that the client has lost the bond,
   enabling encryption will fail.'

Fixes #17983

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

### CODE REVIEW

**Review by carlescufi — COMMENTED** [review]
Looks pretty good to me, thanks @Vudentz. Don't we need a check in `bt_gatt_notify()` though? if the application passes a `NULL` pointer we still need to check the security level at that point, or is that done in the connection callback?

**Comment by Vudentz** [comment]
> Looks pretty good to me, thanks @Vudentz. Don't we need a check in `bt_gatt_notify()` though? if the application passes a `NULL` pointer we still need to check the security level at that point, or is that done in the connection callback?

The callback will check the value set in the user_data (bt_gatt_ccc), that will be set to 0x00 (disabled) in case it could not be restored due to security level, to restored CCC which requires security bt_gatt_encrypt_change shall be called.

**Comment by nvlsianpu** [comment]
cc @rakons

**Comment by Vudentz** [comment]
@jhedberg @carlescufi It seems we have forgotten this one.

File: subsys/bluetooth/host/gatt.c [hunk 1]
```diff
@@ -1735,6 +1741,25 @@ static u8_t connected_cb(const struct bt_gatt_attr *attr, void *user_data)
...
+			bt_security_t sec = BT_SECURITY_L2;
+
+			if (attr->perm & BT_GATT_PERM_WRITE_AUTHEN) {
+				sec = BT_SECURITY_L3;
+			}
+
+			/* Check if current security is enough */
+			if (conn->sec_level < sec) {
+				if (data->sec < sec)
+					data->sec = sec;
```
Comment (jhedberg) [inline subsys/bluetooth/host/gatt.c]: Missing `{ }`

**Review by jhedberg — APPROVED** [review]
Looks fine except for one coding style issue (missing braces)

File: subsys/bluetooth/host/gatt.c (hunk 1 above)
Comment (Vudentz) [reply subsys/bluetooth/host/gatt.c]: Fixed.

**Comment by jhedberg** [comment]
@carlescufi @joerchan could you please review?

**Comment by carlescufi** [comment]
@Vudentz @jhedberg yes, apologies for the delay on this one. We have had holidays and travelling. We will review this week.

File: subsys/bluetooth/host/gatt.c [hunk 2]
```diff
@@ -1735,6 +1741,26 @@ static u8_t connected_cb(const struct bt_gatt_attr *attr, void *user_data)
 			continue;
 		}
 
+#if defined(CONFIG_BT_SMP)
+		/* Check if attribute requires encryption/authentication */
+		if (attr->perm & BT_GATT_PERM_WRITE_ENCRYPT ||
```
Comment (nvlsianpu) [inline subsys/bluetooth/host/gatt.c]:
```suggestion
		if (attr->perm &( BT_GATT_PERM_WRITE_ENCRYPT | BT_GATT_PERM_WRITE_AUTHEN)) {
```
reason: less assembly instruction in most cases

**Review by nvlsianpu — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/host/gatt.c (hunk 2 above)
Comment (Vudentz) [reply subsys/bluetooth/host/gatt.c]: Indeed that would be better, will send an update shortly.

**Comment by nvlsianpu** [comment]
cc @pdunaj

File: subsys/bluetooth/host/gatt.c (hunk 2 above)
Comment (Vudentz) [reply subsys/bluetooth/host/gatt.c]: I did the update but found out we could reuse the logic for checking the attribute permission for the CCC, so I end up adding another patch.

File: subsys/bluetooth/host/gatt_internal.h [hunk 3]
```diff
@@ -11,8 +11,20 @@
 #define BT_GATT_CENTRAL_ADDR_RES_NOT_SUPP	0
 #define BT_GATT_CENTRAL_ADDR_RES_SUPP		1
 
+#define BT_GATT_PERM_READ_MASK			(BT_GATT_PERM_READ | \
```
[the remark this answers is not in the document]
Comment (Vudentz) [reply subsys/bluetooth/host/gatt_internal.h]: This is not a public API, it is only to be used with bt_gatt_check_perm which is used by att.c as well.

[the remark this answers is not in the document]
Comment (Vudentz) [reply subsys/bluetooth/host/gatt_internal.h]: Id like this to be merge at some point so we need to be clear if this requires another change or it is fine to be applied.

[the remark this answers is not in the document]
Comment (nvlsianpu) [reply subsys/bluetooth/host/gatt_internal.h]: Ach right - I missed _MASK at the ends

**Review by nvlsianpu — APPROVED** [review]

**Comment by jhedberg** [comment]
@Vudentz the shippable failure seems valid. Didn't check all failures but this one is at least there:
```
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/bluetooth/host/gatt.c: In function 'update_ccc':
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/bluetooth/host/gatt.c:1767:7: error: unused variable 'err' [-Werror=unused-variable]
  u8_t err;
       ^~~
```

**Comment by Vudentz** [comment]
> @Vudentz the shippable failure seems valid. Didn't check all failures but this one is at least there:
> 
> ```
> /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/bluetooth/host/gatt.c: In function 'update_ccc':
> /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/subsys/bluetooth/host/gatt.c:1767:7: error: unused variable 'err' [-Werror=unused-variable]
>   u8_t err;
>        ^~~
> ```

Yep, Ive just fixed it.

File: samples/bluetooth/peripheral/src/main.c [hunk 4]
```diff
@@ -194,7 +194,8 @@ BT_GATT_SERVICE_DEFINE(vnd_svc,
 			       BT_GATT_PERM_READ_ENCRYPT |
 			       BT_GATT_PERM_WRITE_ENCRYPT,
 			       read_vnd, write_vnd, vnd_value),
-	BT_GATT_CCC(vnd_ccc_cfg_changed),
+	BT_GATT_CCC(vnd_ccc_cfg_changed,
+		    BT_GATT_PERM_READ | BT_GATT_PERM_WRITE_ENCRYPT),
```
[the remark this answers is not in the document]
Comment (joerchan) [reply samples/bluetooth/peripheral/src/main.c]: Is this a requirement of the VND service?

File: include/bluetooth/conn.h [hunk 5]
```diff
@@ -334,6 +334,12 @@ typedef enum __packed {
  */
 int bt_conn_set_security(struct bt_conn *conn, bt_security_t sec);
 
+/** @brief Get security level for a connection.
+ *
+ *  @return Connection security level
+ */
+bt_security_t bt_conn_get_security(struct bt_conn *conn);
```
Comment (joerchan) [inline include/bluetooth/conn.h]: Maybe move below the __deprecated declaration of the old bt_conn_security?

File: subsys/bluetooth/host/conn.c [hunk 6]
```diff
@@ -1085,6 +1085,16 @@ int bt_conn_set_security(struct bt_conn *conn, bt_security_t sec)
...
 }
+
+bt_security_t bt_conn_get_security(struct bt_conn *conn)
+{
+	return conn->sec_level;
+}
+#else
+bt_security_t bt_conn_get_security(struct bt_conn *conn)
+{
+	return BT_SECURITY_L1;
```
Comment (joerchan) [inline subsys/bluetooth/host/conn.c]: Should it be L0 for BR/EDR? (It looks like conn->sec_level is not initialized for BR/EDR connections, see `bt_conn_add_br`)

**Review by joerchan — APPROVED** [review]
Looks good. Just some minor questions.

File: include/bluetooth/conn.h (hunk 5 above)
Comment (Vudentz) [reply include/bluetooth/conn.h]: For what reason?

File: samples/bluetooth/peripheral/src/main.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (Vudentz) [reply samples/bluetooth/peripheral/src/main.c]: Yes it follows the chrc permissions.

File: subsys/bluetooth/host/conn.c (hunk 6 above)
Comment (Vudentz) [reply subsys/bluetooth/host/conn.c]: Well L0 is only used for SDP iirc.

**Comment by Vudentz** [comment]
@jhedberg can we finally gets this merged?

**Comment by jhedberg** [comment]
> @jhedberg can we finally gets this merged?

@Vudentz done!

**Pull request merged**